### PR TITLE
Barcode: lookup item

### LINKED ISSUE #1524: Barcoding: lookup item
## Is your feature request related to a problem? Please describe 👀
These lines from requirements:
"Add API method to lookup an item by barcode, or add all barcodes to the item query"
"Update OS detail view, to add item by barcode"
"Lookup a stock line by batch and expiry date; prompt the user for the quantity to supply"

## Describe the solution you'd like 🎁
We currently have all items locally. A filter would be useful, so this could be added too - to meet the requirement, it would be enough to add the barcode to the item query.

There's a button to initiate a scan, ideally though we should add a button to start auto scanning.

Tasks
- [ ] action to start and stop auto scan
- [ ] button on outbound shipment to add via auto scan
- [ ] on a successful scan result, open the 'add item' modal with the item selected
- [ ] focus on the quantity for the batch, as identified by the barcode?

### Moneyworks Jobcode 🧰

`DJUNCUST`

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1524

# 👩🏻‍💻 What does this PR do? 
 
Allow a user to scan items when populating an outbound shipment.

- [ ] Add a button to outbounds ( if a scanner is found ) which starts and stops auto scan
- [ ] On a successful scan result, open the 'add item' modal with the item selected
- [ ] On an unsuccessful result - show a warning and open the add item modal without selecting an item 

There's another issue to come which will hook up the android code, save barcodes back and highlight specific batches.

This has required more work than expected; some of the extra work is:
- adding a barcode lookup query, which allows filtering by the code
- refactoring the barcode scanning to allow for callback events, this is so that the user can enable scanning and keep scanning / adding items

# 🧪 How has/should this change been tested? 

You'll need to start electron. For development `yarn electron:start-local` (after running the server and running 'yarn start-local') or more usually `yarn electron:start` though for that, you'll have to build f/e and then start your server, in order to pickup all the f/e changes here.

You can then
- open an outbound; check that there's a scan button
- when clicking scan, it changes the icon and the text to 'stop'
- clicking 'stop' will stop
- navigating away from the outbound will also stop ( save, cancel, reload, and navigation )
- when scanning, a scan result will pop up the add item window
  - with the item selected, if there's a match ( you'll need the barcode in your barcode table )
  - with the item autocomplete open otherwise
- the edit stock line functionality, where you can scan barcodes, should work as normal
- Open the web version, and check that there's no scan button

To manually add a barcode, you can scan an item in the edit-stock-line modal, copy the result and add an entry to your barcode table. You'll need an item id, and will have to copy the barcode value to the `value` column.

You can then check that using the graphQL api too.
here's how it looks:

Screenshot 2023-04-18 at 3 25 53 PM

https://user-images.githubusercontent.com/9192912/232927445-d2429e9b-e683-4ca0-a22e-17ae343c4f95.mp4

## 💌 Any notes for the reviewer?

There are two USB scanners in the office - you'll need one to test fully
To make the continuous scan work, I enabled on the scanner:

image

and they also recommend enabling picklist mode ("In this mode, you can pick out and decode a bar code from a group of bar codes that are printed close together by placing the aiming pattern on the bar code you want to decode.")

image

## 📃 Documentation
- [ ] Barcode scanning in outbounds

### CODE REVIEW

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (jmbrunskill) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:49]: If the intention is for a one to one match between a `barcode` and a `barcode:value` I wonder if it would be nicer to had a `getBarcode` endpoint, rather than using the filters and `getBarcodes` endpoint in this context.
Obviously a bit more work so I'm ok with how things are. Just think that would communicate more clearly the expectation.

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (jmbrunskill) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:44]: Where would it save to? Invoice Line?

**Review by jmbrunskill — COMMENTED** [review]
Looks good from a quick glance at the code.
Haven't tested anything...

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (mark-prins) [reply client/packages/common/src/utils/BarcodeScannerContext.tsx:36]: it's popping up now that I've enabled continuous mode - seems to be a sync idle character, according to the ASCII reference. A little annoying.. without stripping it out, then the parsing fails

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:44]: Just saving back to the barcode table, against the selected item_id. This allows lookup for next time

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:49]: true. I was wondering at one point if we'd have matching items come back, along with pack sizes. We could then do something with the pack size info.. now that I've done the rest of it.. we aren't using the pack sizes anyway

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (andreievg) [reply client/packages/common/src/utils/BarcodeScannerContext.tsx:36]: I think James suggesting a comment

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (andreievg) [inline client/packages/common/src/utils/BarcodeScannerContext.tsx:98]: Is this try catch statement correct ? (promise should ether resolve or reject not both), if try catch was wrapping `electronNativeAPI.onBarcodeScan` it would make a little bit more sense, although I think this should just resolve(data) without reject ?

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:44]: Looks like this might be done in the next PR ?

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:26]: Hmm, wouldn't this use `isOutboundDisabled` ?

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:72]: Should this be here ?

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:27]: Do we need to add from master list ?

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:29]: I think this is also master list related ?

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:64]: Nice

File: client/packages/invoices/src/OutboundShipment/api/api.ts
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/api/api.ts:159]: `Promise<BarcodesQuery["barcodes"]>` -> would also mean if we update query we don't need to change things here

File: client/packages/invoices/src/OutboundShipment/api/operations.generated.ts
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/api/operations.generated.ts:366]: i like the idea of just querying for what's needed/used (one advantage of graphql)

Comment (andreievg) [reply client/packages/invoices/src/OutboundShipment/api/operations.generated.ts:366]: oops this is meant to be for client/packages/invoices/src/OutboundShipment/api/operations.graphql

File: server/graphql/general/src/queries/barcode.rs
Comment (andreievg) [inline server/graphql/general/src/queries/barcode.rs:20]: I've noticed we've used `value` for barcode field and filter, but for sort it's Barcode, I kind of like barcode instead of value tbh, either way, if it's value or barcode would be great to have it consistent

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:49]: I think eventually barcodes API would be useful, but yeah I agree, with James, I like the idea of consumer driven API development, i.e. first do the front end with mock API, then fill API with functionality

**Review by andreievg — APPROVED** [review]
Looks good to approve (although I would typically test, but I think it's fine to do overall test, see below). There were a few question marks and some improvement suggestions. I think it's best if they are done (if you agree with them) in downstream PR or in a new issue/PR

> ( if a scanner is found )

I wonder what our UX strategy for this is (if not configured or not found or error etc.. do we just hide the button ? "the less you know the better you sleep" hehe )

I had to look a little bit deeper to figure out a couple of things and saw `parseBarcodeData` i didn't look at it's implementation but though it would be great to have some test and test data for it (would be a a great way to explain how it works, with input and output)

`if (!hasElectronApi) error(t('error.unable-to-read-barcode'))();`

Does it mean it will do error only if it's not electron API ?

Just wondering if scan is ongoing, and then it's cancelled (by timeout), what happens to the `.await()` (since stopping scanning should stop all of the events bound with `.on`) ? I was wondering if we can keep previous .read for single scan, it seems it would be simpler (and maybe differentiate by method name `starScanning`, `stopScanning`, `singleScan`)

I want to test all of the above but would want to wait for all of the barcoding to be merged and maybe documentation to be made, I kind of got confused by the `Any notes for the reviewer` section and barcodes and different modes of operation there (I would assume this would be very clear in the docs)

**Comment by GaryWilletts** [comment]
@mark-prins not sure whether this is the place but I wanted to draw your attention to 2 issues I put on the mSupply OG repo for barcodes following my latest snooping when I'm in warehouses!:

https://github.com/sussol/msupply/issues/10291 - GTIN being shown in a separate barcode from other details (very important to accommodate I think, should be very simple). On cartons, not primary packaging so might not be important right now.
https://github.com/sussol/msupply/issues/10292 - element 30 appearing in carton barcodes (very helpful but probably not relevant to what you're doing here)

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (mark-prins) [reply client/packages/common/src/utils/BarcodeScannerContext.tsx:36]: and I was thinking that it wasn't needed, partly with the rework of the function and partly because I don't know why it's actually happening. But sure, will add something fwiw

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:49]: on the other hand, if we only ever implement the bare minimum which is closely aligned to the current implementation of the front end then we commit ourselves to api development for any future change. The advantage of graphQL is partly that we have features available and can quickly modify the f/e when implementing functionality.

Given that this is the consenus, I'll strip the functionality out: added to #1629

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:44]: yeah, is in the next PR

File: client/packages/common/src/utils/BarcodeScannerContext.tsx
Comment (mark-prins) [reply client/packages/common/src/utils/BarcodeScannerContext.tsx:98]: It seemed like a neat way to handle the error, but I agree. Shouldn't be trying to resolve and reject. I'll remove in #1629

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:26]: I thought that it was only for new shipments, I guess we can still add to picked and allocated - it's just not the usual workflow. will update, ta.

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:27]: oops

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (mark-prins) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:72]: oops

File: server/graphql/general/src/queries/barcode.rs
Comment (mark-prins) [reply server/graphql/general/src/queries/barcode.rs:20]: I was resisting using `barcode.barcode` as per mSupply implementation. `value` isn't great, but it is less repetitive.
Might review that though in #1629

**Comment by mark-prins** [comment]
thanks for reviews @andreievg & @jmbrunskill! much appreciated.
Follow up in #1629 - rather than forcing a re-review, since there's functional changes requested.

Thanks @GaryWilletts for those - will review. The package we're using will pick out those elements of the barcodes, so we should decide what to do with them

**Comment by GaryWilletts** [comment]
@mark-prins Cool. I assume the "package we're using" will recognise all GS1 elements? In mSupply OG we use a spreadsheet, provided on the GS1 website, to import the elements that can appear in a GS1 barcode. We do the update occasionally, although it was a long time ago when we last did it! When I last looked the GS1 definitions spreadsheet was updated surprisingly regularly.

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx
Comment (andreievg) [reply client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:49]: > The advantage of graphQL is partly that we have features available

This can also be done with rest, not sure it's specific to graphQL

> if we only ever implement the bare minimum which is closely aligned to the current implementation of the front end then we commit ourselves to api development for any future change

Why do we `commit` ? Api can be changed and extended, and with shared types between front and back end so it's safe to do. 

I am still very strong about API based on current consumer requirements (rather then guessing what requirements should be and implementing something that prioritises future requirements vs ease of consumption of current requirements)

**Comment by mark-prins** [comment]
@GaryWilletts it does currently - though if the GS1 definitions are changing regularly then perhaps we'll be out of date. The last publish of [gs1-barcode-parser-mod](https://www.npmjs.com/package/gs1-barcode-parser-mod) was a year ago now!

**Pull request merged**
